llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -136,13 +140,115 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorShiftsByZero(
   return true;
 }
 
+bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  MachineInstr *ExtOp = getDefIgnoringCopies(MI.getOperand(2).getReg(), MRI);
+
+  if (!isa<GIntrinsic>(MI) || !isa<GIntrinsic>(*ExtOp))
+    return false;
+  switch (cast<GIntrinsic>(MI).getIntrinsicID()) {
+  case Intrinsic::aie2_set_I512_I128: {
+    if (cast<GIntrinsic>(*ExtOp).getIntrinsicID() !=
+        Intrinsic::aie2_extract_I128_I512)
+      return false;
+    break;
+  }
+  case Intrinsic::aie2_set_I512_I256: {
+    const Register IdxReg = MI.getOperand(3).getReg();
+    if (cast<GIntrinsic>(*ExtOp).getIntrinsicID() !=
+            Intrinsic::aie2_ext_I256_I512 ||
+        IdxReg != ExtOp->getOperand(3).getReg())
```

Comment:
I suggest to include one more case here: registers are different but they trace back to the equal G_CONSTANT. Maybe this can be the only rule here. I saw cases where we have an exclusive register for each intrinsic, so it can be more general.